File: ios/Packages/DesignSystem/Sources/DesignSystem/Tokens/ImageType.swift
import SwiftUI

/// Kind of image surface. Bundles the size-class-keyed width (``width(_:)``) with the matching
/// aspect ratio (``aspect``) so a call site declares the kind once instead of pairing an
/// `ImageDimens` width with an `ImageDimens` aspect by hand. This is layout sizing only; image
/// fetching/bucket selection is a separate concern handled by `TmdbImageType`.
public enum ImageType {
    case poster
    case backdrop
    case cast

    /// Aspect ratio for the kind, expressed as width / height (so `height = width / aspect`).
    public var aspect: CGFloat {
        switch self {
        case .poster: ImageDimens.posterAspect
        case .backdrop: ImageDimens.backdropAspect
        case .cast: ImageDimens.castAspect
        }
    }

    /// Width for the kind at the given window width size class.
    public func width(_ sizeClass: WidthSizeClass) -> CGFloat {
        switch self {
        case .poster: ImageDimens.posterWidth(sizeClass)
        case .backdrop: ImageDimens.backdropCardWidth(sizeClass)
        case .cast: ImageDimens.castCardWidth(sizeClass)
        }
    }
}

File: android-designsystem/src/main/kotlin/com/thomaskioko/tvmaniac/compose/theme/ImageType.kt
package com.thomaskioko.tvmaniac.compose.theme

import androidx.compose.runtime.Composable
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.ui.unit.Dp

/**
 * Kind of image surface. Bundles the size-class-keyed [width] with the matching [aspect] so a call
 * site declares the kind once instead of pairing a [Layout] width with an [ImageDimens] aspect by
 * hand. Layout sizing only; image fetching/bucket selection stays with the TMDB interceptor.
 */
public enum class ImageType {
    Poster,
    Backdrop,
    Cast,
    ;

    /** Aspect ratio for the kind, expressed as width / height (matching `Modifier.aspectRatio`). */
    public val aspect: Float
        get() = when (this) {
            Poster -> ImageDimens.PosterAspect
            Backdrop -> ImageDimens.BackdropAspect
            Cast -> ImageDimens.CastAspect
        }

    /** Width for the kind at the current window width size class. */
    public val width: Dp
        @Composable @ReadOnlyComposable
        get() = when (this) {
            Poster -> Layout.posterWidth
            Backdrop -> Layout.backdropCardWidth
            Cast -> Layout.castCardWidth
        }
}
